### PR TITLE
feat: customization for daemon deployment. Closes #223

### LINKED ISSUE #223: Customization for daemon service
# Summary

We need to have a way to customize generated daemon service deployment (such as replicas, resources, env for log level, etc).

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by dseapy**
Signed-off-by: David Seapy <dseapy@gmail.com>

Closes #223.

Adds customization for daemon deployment.

### CODE REVIEW

File: pkg/apis/numaflow/v1alpha1/pipeline_types.go
Comment (vigith) [inline pkg/apis/numaflow/v1alpha1/pipeline_types.go:398]: thinking whether we nest it under `LifeCycle`, this is because we have to also add Job resources and those jobs are for setting up and tearing down the pipeline. The Daemon Server is also used to manage the lifecycle too, e.g., pausing the pipeline, etc.

**Review by vigith — COMMENTED** [review]
I think we should document this, perhaps create `Lifecycle` in https://numaflow.numaproj.io/pipeline/ and nest Daemon details under it?

File: pkg/apis/numaflow/v1alpha1/pipeline_types.go
Comment (vigith) [reply pkg/apis/numaflow/v1alpha1/pipeline_types.go:398]: option (c) sounds good to me, followed by (b).

Comment (dseapy) [reply pkg/apis/numaflow/v1alpha1/pipeline_types.go:398]: sounds good.  should both `Templates` and `DaemonTemplate` be pointers?

Comment (dseapy) [reply pkg/apis/numaflow/v1alpha1/pipeline_types.go:398]: Also, for my future understanding, what goes into the decision when choosing to use the pointer vs not for structs?  I see Lifecycle and Watermark aren't but PipelineLimits is.

Is it that in some cases we want to distinguish between nil and the empty value?

**Comment by dseapy** [comment]
> I think we should document this, perhaps create `Lifecycle` in https://numaflow.numaproj.io/pipeline/ and nest Daemon details under it?

* are you saying have a `lifecycle.md` alongside the `watermarks.md` and have a section in that markdown for Daemon configuration?
* if using `Templates` in the Pipeline Spec and `DaemonTemplate` (option c), then does that change where the documentation should go, or does it still belong in lifecycle docs

File: pkg/apis/numaflow/v1alpha1/pipeline_types.go
Comment (whynowy) [reply pkg/apis/numaflow/v1alpha1/pipeline_types.go:398]: I think both `Templates` and `DaemonTemplate` should be pointers.

My consideration on pointers or structs are:
1. When the property is missing in the spec, after applying to k8s, do you want the users to see it with `kubectl get pl xxx -o yaml`. I think we don't want the users to see `templates` and `pipelineLimits` if you don't care about them.
2. Is there any nested properties that always having values, if yes, structs will make things easier (no need to do nil check). For example, `watermark.disabled`, `lifecycle.desiredPhase`.

**Comment by whynowy** [comment]
> > I think we should document this, perhaps create `Lifecycle` in https://numaflow.numaproj.io/pipeline/ and nest Daemon details under it?
> 
> * are you saying have a `lifecycle.md` alongside the `watermarks.md` and have a section in that markdown for Daemon configuration?
> * if using `Templates` in the Pipeline Spec and `DaemonTemplate` (option c), then does that change where the documentation should go, or does it still belong in lifecycle docs

Adding a page next to `Init Containers` named something like `Pipeline Customization` (find a better name?) to describe the pod customization for daemon and jobs? And add links in `Container Resources` and `Environment Variables` to this page?

File: pkg/apis/numaflow/v1alpha1/pipeline_types.go
Comment (whynowy) [reply pkg/apis/numaflow/v1alpha1/pipeline_types.go:422]: `daemon` is okay since it's already in `templates`, 👍

**Comment by dseapy** [comment]
> Adding a page next to `Init Containers` named something like `Pipeline Customization` (find a better name?) to describe the pod customization for daemon and jobs?

I can't think of a better name, so went with that for now.  can change if ya'll think of something.

File: docs/pipeline-customization.md
Comment (whynowy) [inline docs/pipeline-customization.md:24]: Missing `metadata`

**Review by vigith — APPROVED** [review]

**Comment by dseapy** [comment]
i've got a branch based on this one that does the JobTemplate stuff, was going to put that up for PR after this merges, but i'm seeing https://github.com/numaproj/numaflow/pull/296, so maybe best i hold off on that one until the refactor goes through?

**Pull request merged**
